Math/NumberTheory/Zeta/Dirichlet.hs

```diff
@@ -0,0 +1,210 @@
+-- |
+-- Module:      Math.NumberTheory.Zeta.Dirichlet
+-- Copyright:   (c) 2018 Alexandre Rodrigues Baldé
+-- Licence:     MIT
+-- Maintainer:  Alexandre Rodrigues Baldé <alexandrer_b@outlook.com>
+-- Stability:   Provisional
+-- Portability: Non-portable (GHC extensions)
+--
+-- Dirichlet beta-function.
+
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Math.NumberTheory.Zeta.Dirichlet
+  ( betas
+  , betasEven
+  , betasOdd
+  , euler
+  , euler'
+  , eulerPolyAt1
+  ) where
+
+import Data.ExactPi                     (ExactPi (..), approximateValue)
+import Data.List                        (zipWith4)
+import Data.Ratio                       (Ratio, (%), numerator)
+
+import Math.NumberTheory.Recurrencies   (factorial, stirling2)
+import Math.NumberTheory.Zeta.Riemann   (zetasOdd)
+import Math.NumberTheory.Zeta.Utils     (intertwine, skipEvens, skipOdds,
+                                         suminf)
+
+-- | Infinite zero-based list of <https://en.wikipedia.org/wiki/Euler_number Euler numbers>.
+-- The algorithm used was derived from <http://www.emis.ams.org/journals/JIS/VOL4/CHEN/AlgBE2.pdf Algorithms for Bernoulli numbers and Euler numbers>
+-- by Kwang-Wu Chen, second formula of the Corollary in page 7.
+--
+-- >>> take 10 euler :: [Rational]
+-- [1 % 1,0 % 1,(-1) % 1,0 % 1,5 % 1,0 % 1,(-61) % 1,0 % 1,1385 % 1,0 % 1]
+euler :: forall a . Integral a => [Ratio a]
+euler = map (f . tail) (tail stirling2)
+  where
+    f = sum . zipWith4 (\sgn fact a stir -> sgn * fact * stir * a) (cycle [1, -1])
+                                                                   factorial
+                                                                   as
+
+    as :: Integral a => [Ratio a]
+    as = zipWith3 (\sgn frac ones -> (sgn * ones) % frac) (cycle [1, 1, 1, 1, -1, -1, -1, -1])
+                                                          (dups (iterate (2 *) 1))
+                                                          (cycle [1, 1, 1, 0])
```

Comment:
Sorry for bike shedding, but these hanging expressions do not look good to me. I'd rather put arguments of `zipWith3` on the next lines, with one more level of indentation.